### ISSUE TITLE
ReduceScatter autograd (jvp/vjp/vmap)

### ISSUE THREAD

**Issue by Goekdeniz-Guelmez** [ml-explore/mlx#4405]
currently distributed.sum_scatter has no autograd (vjp/jvp), does a PR exist or should I do one?

I need this for building FSDP/ZeRO-3-style training on top of `mx.distributed`, I noticed that `sum_scatter` (ReduceScatter primitive) defines no `jvp/vjp/vmap`  and mlx/distributed/ primitives.cpp only implements autograd for AllReduce and AllGather. Calling `mx.grad` through a graph that contains `sum_scatter` therefore fails.

- `vjp` of sum-scatter (Sum) = `all_gather` of the cotangent along axis 0
- `jvp` = `sum_scatter` of the tangent
- `vmap` = re-issue `sum_scatter`, pass axes through

Current workaround on my side is an `mx.custom_function` wrapper, but in-graph support would let `mx.compile` fuse it.

☑️ I understand it is strictly prohibited to use AI to write issues.

**Comment by zcbenz** [comment]
I'm not sure, @nastya236 should have a better idea. But I have to say that this is the type of contribution that we would be very reluctant to take from external, because verifying the correctness basically takes as much time as writing it ourselves and we usually are not able to find an apple employee to do that.

**Comment by nastya236** [comment]
Thanks for the interest. 
We already have FSDP implemented https://github.com/ml-explore/mlx/pull/3768 exactly with a custom function. Also based on my understanding of what you are trying to implement, I am not sure that you need sum_scatter vjp, because for the forward you gather parameters, for the backward you sum_scatter gradients. Regarding, mx.compile fusion I am not sure that I fully understand fusion you are talking about.

**Comment by Goekdeniz-Guelmez** [comment]
Thanks a lot @nastya236 ,

the custom function approach should cover the basic parameter all-gather - gradient sum-scatter FSDP path.

My intended use is more. Full MLX-LM-LoRA training (both full fine-tuning and LoRA) with TP plus FSDP/ZeRO-style sharding, including reduce-scattered activations between blocks. sum_scatter appeers in the normal forward graph, so MLX needs to differentiate through it, VJP should be all_gather on the cotangent.

Would a focused PR for sum_scatter’s be welcomed?

**Comment by nastya236** [comment]
TP is implemented in distributed layers : https://github.com/ml-explore/mlx/blob/main/python/mlx/nn/layers/distributed.py
We shard weights and activations in reduction dimension, so all_reduce is sufficient for both forward and backward, no need for sum_scatter.
Regarding pull request, thanks for the interest, unfortunately I think I would be more comfortable doing such a change myself as Cheng mentioned, because verifying the correctness takes as much time as writing it ourselves.. But for now this is not a priority.

**Comment by Goekdeniz-Guelmez** [comment]
yea, that makes sense. I’ll use the existing custom-function FSDP path for real distributed training and MLX’s distributed layers for tensor parallelism. Thanks for the direction.
